### PR TITLE
Resolve issues with foreground tasks getting cancelled

### LINKED ISSUE #205: Long running tasks are cancelled after 10 minutes
## Observed behavior

Long running tasks like a channel resource import are cancelled after 10 minutes.

## Errors and logs

```
Work [ id=31c6913f-9e0b-498c-8a51-b8eb6bec827a, tags={ kolibri_job_type:kolibri.core.content.tasks.remotecontentimport, kolibri_job_long_running, kolibri_task_id:af5648f9761d492fa8c12b282acfcf58, org.learningequality.Kolibri.ForegroundWorker } ] was cancelled
java.util.concurrent.CancellationException: Task was cancelled.
  at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
  at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
  at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
  at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
  at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  at java.lang.Thread.run(Thread.java:923)
```

## Expected behavior

Our long-running tasks should be run in a foreground service, which bypasses the 10 minute limitation and the tasks are not cancelled after 10 minutes.

## User-facing consequences

Large channel imports get interrupted and even if they start over, which is still slow, can make it nearly impossible to complete a full import of a channel.

## Steps to reproduce

Import a large channel that can take over 10 minutes to fully import.

## Context

Kolibri 0.16.0
Kolibri for Android 0.1.0
Android 11

### PR DESCRIPTION
**Pull request by bjester**
## Summary
- Adds dependency `androidx.lifecycle:lifecycle-service:2.7.0` since `SystemForegroundService` relies on it (possible this was the problematic oversight)
- Removes some unused imports
- Adds `PythonProvider` class which implements `AutoCloseable` for use with try-with-resources, to provider the android context to the python side, and automatically ensure we don't create context memory leaks
- Updates `ContextUtil` to utilize the `PythonProvider` if it's active
- Refactors work manager worker logic into abstract base class and makes foreground worker a regular `Worker` instead of `RemoteListenableWorker`
- Re-adds `setForegroundAsync` that was mistakenly deleted, but does not call it repeatedly-- calling it repeatedly seemed to have no benefit against the issue but to spam the logs
- Goes back to having a single android notification for tasks, and since foreground tasks require an integer ID, it converts the task request UUID to an integer using CRC32
- Removes `WorkerService` which was a `RemoteWorkerService` for running `RemoteListenableWorker`s
- Adds observer pattern to task worker implementation so task update calls go through it, which is a Java object that should only exist in the same task's thread
- Adds `onProgressUpdate` hook method to worker base class which is called when the task's progress is updated, and by default calls `setProgressAsync` and updates the notification (this may be the true fix)
- Fixes order of operation issue after switching to `//` division instead of using `math.floor(...)` changed in https://github.com/learningequality/kolibri-installer-android/issues/204
- Adds `android.permission.POST_NOTIFICATIONS` permission and check for it since Android Studio was complaining about it missing (I don't recall this happening previously)

## References
Fixes https://github.com/learningequality/kolibri-installer-android/issues/205

### CODE REVIEW

**Review by radinamatic — APPROVED** [review]
APK tested on an Android 11 tablet, 6GBs from the CK12 channel imported without a single hickup in the logcat or UI 👏🏽 💯 :shipit: 

logcat | task completed
-- | --
[Android11-TabA7.log](https://github.com/learningequality/kolibri-installer-android/files/14672531/Android11-TabA7.log) |

**Comment by rtibbles** [comment]
> Goes back to having a single android notification for tasks, and since foreground tasks require an integer ID, it converts the task request UUID to an integer using CRC32

Nice!

File: python-for-android/dists/kolibri/build.gradle [hunk 1]
```diff
@@ -108,5 +108,6 @@ dependencies {
     implementation 'androidx.concurrent:concurrent-futures:1.1.0'
     implementation 'androidx.work:work-runtime:2.9.0'
     implementation 'androidx.work:work-multiprocess:2.9.0'
+    implementation "androidx.lifecycle:lifecycle-service:2.7.0"
```
Comment (rtibbles) [inline python-for-android/dists/kolibri/build.gradle:111]: Did Android Studio warn you about this?

File: src/android_app_plugin/kolibri_plugin.py [hunk 2]
```diff
@@ -73,11 +74,10 @@ def update(self, job, orm_job, state=None, **kwargs):
             # avoid passing integers that are too large
             # PROGRESS_LIMIT gives sufficient precision for a % progress calculation
             if total_progress > PROGRESS_LIMIT:
-                progress = progress // total_progress * PROGRESS_LIMIT
+                progress = PROGRESS_LIMIT * progress // total_progress
```
Comment (rtibbles) [inline src/android_app_plugin/kolibri_plugin.py:77]: Oh yes... the previous version would always go to 0... I guess the math didn't check out!

**Review by rtibbles — APPROVED** [review]
Thanks to the PR description, I understand this at a high level - although don't feel completely equipped to comment on specific changes. But it also works in manual testing, and feels like a decent amount of cleanup and consolidation too!

File: python-for-android/dists/kolibri/build.gradle (hunk 1 above)
Comment (bjester) [reply python-for-android/dists/kolibri/build.gradle:111]: It did not! :facepalm: Besides highlighting the inherited classes as red (missing), which still required you to click through and look at the class itself.

File: src/android_app_plugin/kolibri_plugin.py (hunk 2 above)
Comment (bjester) [reply src/android_app_plugin/kolibri_plugin.py:77]: I originally had `math.floor(...)` and tested it with that, then I thought "oh yeah there's integer division `//` which is simpler" (self-sabotage)

**Pull request merged**